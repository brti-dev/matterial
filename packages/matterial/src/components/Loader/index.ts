export * from './loader'
export { default } from './loader'
