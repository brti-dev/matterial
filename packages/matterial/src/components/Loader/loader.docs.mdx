---
name: Loader
description: Indicates something's happening.
---

## Usage

```jsx
import { Loader } from 'matterial'

function App() {
  return <Loader />
}
```

## Loader Props

### Size

`size = number`

<Container row>
  <Loader size={20} />
  <Loader />
  <Loader size={100} />
</Container>

```jsx
<Loader size={20} />
<Loader />
<Loader size={100} />
```

### Color

`color = string | Color`

<Container row>
  <Loader color="primary" />
  <Loader color="secondary" />
  <Loader color="accent-3" />
  <Loader color="hotpink" />
</Container>

```jsx
<Loader color="primary" />
<Loader color="secondary" />
<Loader color="accent-3" />
<Loader color="hotpink" />
```
