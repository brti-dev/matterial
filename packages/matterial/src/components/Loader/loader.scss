// Inspired by Vercel Design Spinner https://vercel.com/design/spinner

.loader__container,
.loader {
  width: var(--size, 30px);
  height: var(--size, 30px);
}

.loader {
  position: relative;
  top: 50%;
  left: 50%;
}

.loader__bar {
  -webkit-animation: loader_spin 1.2s linear infinite;
  animation: loader_spin 1.2s linear infinite;
  background: var(--color, var(--color-accent-7));
  border-radius: var(--radius);
  height: 8%;
  left: -10%;
  position: absolute;
  top: -3.9%;
  width: 24%;

  &:first-child {
    -webkit-animation-delay: -1.2s;
    animation-delay: -1.2s;
    transform: rotate(0.0001deg) translate(146%);
  }
  &:nth-child(2) {
    -webkit-animation-delay: -1.1s;
    animation-delay: -1.1s;
    transform: rotate(30deg) translate(146%);
  }
  &:nth-child(3) {
    -webkit-animation-delay: -1s;
    animation-delay: -1s;
    transform: rotate(60deg) translate(146%);
  }
  &:nth-child(4) {
    -webkit-animation-delay: -0.9s;
    animation-delay: -0.9s;
    transform: rotate(90deg) translate(146%);
  }
  &:nth-child(5) {
    -webkit-animation-delay: -0.8s;
    animation-delay: -0.8s;
    transform: rotate(120deg) translate(146%);
  }
  &:nth-child(6) {
    -webkit-animation-delay: -0.7s;
    animation-delay: -0.7s;
    transform: rotate(150deg) translate(146%);
  }
  &:nth-child(7) {
    -webkit-animation-delay: -0.6s;
    animation-delay: -0.6s;
    transform: rotate(180deg) translate(146%);
  }
  &:nth-child(8) {
    -webkit-animation-delay: -0.5s;
    animation-delay: -0.5s;
    transform: rotate(210deg) translate(146%);
  }
  &:nth-child(9) {
    -webkit-animation-delay: -0.4s;
    animation-delay: -0.4s;
    transform: rotate(240deg) translate(146%);
  }
  &:nth-child(10) {
    -webkit-animation-delay: -0.3s;
    animation-delay: -0.3s;
    transform: rotate(270deg) translate(146%);
  }
  &:nth-child(11) {
    -webkit-animation-delay: -0.2s;
    animation-delay: -0.2s;
    transform: rotate(300deg) translate(146%);
  }
  &:nth-child(12) {
    -webkit-animation-delay: -0.1s;
    animation-delay: -0.1s;
    transform: rotate(330deg) translate(146%);
  }
}

@keyframes loader_spin {
  0% {
    opacity: 1;
  }
  to {
    opacity: 0.15;
  }
}
