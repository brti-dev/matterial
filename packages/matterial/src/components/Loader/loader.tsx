import { ColoredElement } from '../../interfaces/theme'
import useColor from '../../lib/use-color'
import './loader.scss'

export interface LoaderProps
  extends React.ComponentPropsWithoutRef<'div'>,
    ColoredElement {
  /**
   * Size in pixels; Default: 30
   */
  size?: number
}

export function Loader({
  color: naturalColor,
  size = 30,
  style: naturalStyle = {},
  ...rest
}: LoaderProps): JSX.Element {
  const color = useColor(naturalColor)

  const style = {
    '--size': `${size}px`,
    ...color.style,
    ...naturalStyle,
  }

  return (
    <div className="loader__container" style={style} {...rest}>
      <div className="loader">
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
        <div className="loader__bar"></div>
      </div>
    </div>
  )
}
