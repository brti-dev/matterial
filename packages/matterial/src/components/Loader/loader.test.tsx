import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'
// import userEvent from '@testing-library/user-event'

// import { render, screen } from '../../../test-utils'
import Loader from './loader'

describe('loader', () => {
  test('should render sizes correctly', () => {
    const tree = renderer
      .create(
        <>
          <Loader />
          <Loader size={50} />
        </>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should render colors correctly', () => {
    const tree = renderer
      .create(
        <>
          <Loader color="primary" />
          <Loader color="accent-1" />
          <Loader color="salmon" />
        </>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
