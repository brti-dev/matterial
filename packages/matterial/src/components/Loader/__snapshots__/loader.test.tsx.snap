// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader should render colors correctly 1`] = `
Array [
  <div
    className="loader__container"
    style={
      Object {
        "--color": "var(--color-primary)",
        "--size": "30px",
      }
    }
  >
    <div
      className="loader"
    >
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
    </div>
  </div>,
  <div
    className="loader__container"
    style={
      Object {
        "--color": "var(--color-accent-1)",
        "--size": "30px",
      }
    }
  >
    <div
      className="loader"
    >
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
    </div>
  </div>,
  <div
    className="loader__container"
    style={
      Object {
        "--color": "salmon",
        "--size": "30px",
      }
    }
  >
    <div
      className="loader"
    >
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
    </div>
  </div>,
]
`;

exports[`loader should render sizes correctly 1`] = `
Array [
  <div
    className="loader__container"
    style={
      Object {
        "--size": "30px",
      }
    }
  >
    <div
      className="loader"
    >
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
    </div>
  </div>,
  <div
    className="loader__container"
    style={
      Object {
        "--size": "50px",
      }
    }
  >
    <div
      className="loader"
    >
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
      <div
        className="loader__bar"
      />
    </div>
  </div>,
]
`;
